File: pathplannerlib/src/main/java/com/pathplanner/lib/commands/PPRamseteCommand.java
package com.pathplanner.lib.commands;

import com.pathplanner.lib.PathPlannerTrajectory;
import com.pathplanner.lib.server.PathPlannerServer;
import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.controller.RamseteController;
import edu.wpi.first.math.controller.SimpleMotorFeedforward;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.math.kinematics.ChassisSpeeds;
import edu.wpi.first.math.kinematics.DifferentialDriveKinematics;
import edu.wpi.first.math.kinematics.DifferentialDriveWheelSpeeds;
import edu.wpi.first.wpilibj.DriverStation;
import edu.wpi.first.wpilibj.Timer;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.*;
import java.util.function.BiConsumer;
import java.util.function.Consumer;
import java.util.function.Supplier;

/** Custom PathPlanner version of RamseteCommand */
public class PPRamseteCommand extends CommandBase {
  private final Timer timer = new Timer();
  private final boolean usePID;
  private final PathPlannerTrajectory trajectory;
  private final Supplier<Pose2d> poseSupplier;
  private final RamseteController controller;
  private final SimpleMotorFeedforward feedforward;
  private final DifferentialDriveKinematics kinematics;
  private final Supplier<DifferentialDriveWheelSpeeds> speedsSupplier;
  private final PIDController leftController;
  private final PIDController rightController;
  private final BiConsumer<Double, Double> output;
  private final boolean useAllianceColor;

  private DifferentialDriveWheelSpeeds prevSpeeds;
  private double prevTime;

  private PathPlannerTrajectory transformedTrajectory;

  private static Consumer<PathPlannerTrajectory> logActiveTrajectory = null;
  private static Consumer<Pose2d> logTargetPose = null;
  private static Consumer<ChassisSpeeds> logSetpoint = null;
  private static BiConsumer<Translation2d, Rotation2d> logError = PPRamseteCommand::defaultLogError;

  /**
   * Constructs a new PPRamseteCommand that, when executed, will follow the provided trajectory. PID
   * control and feedforward are handled internally, and outputs are scaled -12 to 12 representing
   * units of volts.
   *
   * <p>Note: The controller will *not* set the outputVolts to zero upon completion of the path -
   * this is left to the user, since it is not appropriate for paths with nonstationary endstates.
   *
   * @param trajectory The trajectory to follow.
   * @param poseSupplier A function that supplies the robot pose - use one of the odometry classes
   *     to provide this.
   * @param controller The RAMSETE controller used to follow the trajectory.
   * @param feedforward The feedforward to use for the drive.
   * @param kinematics The kinematics for the robot drivetrain.
   * @param speedsSupplier A function that supplies the speeds of the left and right sides of the
   *     robot drive.
   * @param leftController The PIDController for the left side of the robot drive.
   * @param rightController The PIDController for the right side of the robot drive.
   * @param outputVolts A function that consumes the computed left and right outputs (in volts) for
   *     the robot drive.
   * @param useAllianceColor Should the path states be automatically transformed based on alliance
   *     color? In order for this to work properly, you MUST create your path on the blue side of
   *     the field.
   * @param requirements The subsystems to require.
   */
  public PPRamseteCommand(
      PathPlannerTrajectory trajectory,
      Supplier<Pose2d> poseSupplier,
      RamseteController controller,
      SimpleMotorFeedforward feedforward,
      DifferentialDriveKinematics kinematics,
      Supplier<DifferentialDriveWheelSpeeds> speedsSupplier,
      PIDController leftController,
      PIDController rightController,
      BiConsumer<Double, Double> outputVolts,
      boolean useAllianceColor,
      Subsystem... requirements) {
    this.trajectory = trajectory;
    this.poseSupplier = poseSupplier;
    this.controller = controller;
    this.feedforward = feedforward;
    this.kinematics = kinematics;
    this.speedsSupplier = speedsSupplier;
    this.leftController = leftController;
    this.rightController = rightController;
    this.output = outputVolts;
    this.useAllianceColor = useAllianceColor;

    this.usePID = true;

    addRequirements(requirements);

    if (useAllianceColor && trajectory.fromGUI && trajectory.getInitialPose().getX() > 8.27) {
      DriverStation.reportWarning(
          "You have constructed a path following command that will automatically transform path states depending"
              + " on the alliance color, however, it appears this path was created on the red side of the field"
              + " instead of the blue side. This is likely an error.",
          false);
    }
  }

  /**
   * Constructs a new PPRamseteCommand that, when executed, will follow the provided trajectory. PID
   * control and feedforward are handled internally, and outputs are scaled -12 to 12 representing
   * units of volts.
   *
   * <p>Note: The controller will *not* set the outputVolts to zero upon completion of the path -
   * this is left to the user, since it is not appropriate for paths with nonstationary endstates.
   *
   * @param trajectory The trajectory to follow.
   * @param poseSupplier A function that supplies the robot pose - use one of the odometry classes
   *     to provide this.
   * @param controller The RAMSETE controller used to follow the trajectory.
   * @param feedforward The feedforward to use for the drive.
   * @param kinematics The kinematics for the robot drivetrain.
   * @param speedsSupplier A function that supplies the speeds of the left and right sides of the
   *     robot drive.
   * @param leftController The PIDController for the left side of the robot drive.
   * @param rightController The PIDController for the right side of the robot drive.
   * @param outputVolts A function that consumes the computed left and right outputs (in volts) for
   *     the robot drive.
   * @param requirements The subsystems to require.
   */
  public PPRamseteCommand(
      PathPlannerTrajectory trajectory,
      Supplier<Pose2d> poseSupplier,
      RamseteController controller,
      SimpleMotorFeedforward feedforward,
      DifferentialDriveKinematics kinematics,
      Supplier<DifferentialDriveWheelSpeeds> speedsSupplier,
      PIDController leftController,
      PIDController rightController,
      BiConsumer<Double, Double> outputVolts,
      Subsystem... requirements) {
    this(
        trajectory,
        poseSupplier,
        controller,
        feedforward,
        kinematics,
        speedsSupplier,
        leftController,
        rightController,
        outputVolts,
        false,
        requirements);
  }

  /**
   * Constructs a new PPRamseteCommand that, when executed, will follow the provided trajectory.
   * Performs no PID control and calculates no feedforwards; outputs are the raw wheel speeds from
   * the RAMSETE controller, and will need to be converted into a usable form by the user.
   *
   * @param trajectory The trajectory to follow.
   * @param poseSupplier A function that supplies the robot pose - use one of the odometry classes
   *     to provide this.
   * @param controller The RAMSETE follower used to follow the trajectory.
   * @param kinematics The kinematics for the robot drivetrain.
   * @param outputMetersPerSecond A function that consumes the computed left and right wheel speeds.
   * @param useAllianceColor Should the path states be automatically transformed based on alliance
   *     color? In order for this to work properly, you MUST create your path on the blue side of
   *     the field.
   * @param requirements The subsystems to require.
   */
  public PPRamseteCommand(
      PathPlannerTrajectory trajectory,
      Supplier<Pose2d> poseSupplier,
      RamseteController controller,
      DifferentialDriveKinematics kinematics,
      BiConsumer<Double, Double> outputMetersPerSecond,
      boolean useAllianceColor,
      Subsystem... requirements) {
    this.trajectory = trajectory;
    this.poseSupplier = poseSupplier;
    this.controller = controller;
    this.kinematics = kinematics;
    this.output = outputMetersPerSecond;

    this.feedforward = null;
    this.speedsSupplier = null;
    this.leftController = null;
    this.rightController = null;
    this.useAllianceColor = useAllianceColor;

    this.usePID = false;

    addRequirements(requirements);

    if (useAllianceColor && trajectory.fromGUI && trajectory.getInitialPose().getX() > 8.27) {
      DriverStation.reportWarning(
          "You have constructed a path following command that will automatically transform path states depending"
              + " on the alliance color, however, it appears this path was created on the red side of the field"
              + " instead of the blue side. This is likely an error.",
          false);
    }
  }

  /**
   * Constructs a new PPRamseteCommand that, when executed, will follow the provided trajectory.
   * Performs no PID control and calculates no feedforwards; outputs are the raw wheel speeds from
   * the RAMSETE controller, and will need to be converted into a usable form by the user.
   *
   * @param trajectory The trajectory to follow.
   * @param poseSupplier A function that supplies the robot pose - use one of the odometry classes
   *     to provide this.
   * @param controller The RAMSETE follower used to follow the trajectory.
   * @param kinematics The kinematics for the robot drivetrain.
   * @param outputMetersPerSecond A function that consumes the computed left and right wheel speeds.
   * @param requirements The subsystems to require.
   */
  public PPRamseteCommand(
      PathPlannerTrajectory trajectory,
      Supplier<Pose2d> poseSupplier,
      RamseteController controller,
      DifferentialDriveKinematics kinematics,
      BiConsumer<Double, Double> outputMetersPerSecond,
      Subsystem... requirements) {
    this(
        trajectory,
        poseSupplier,
        controller,
        kinematics,
        outputMetersPerSecond,
        false,
        requirements);
  }

  @Override
  public void initialize() {
    if (useAllianceColor && trajectory.fromGUI) {
      transformedTrajectory =
          PathPlannerTrajectory.transformTrajectoryForAlliance(
              trajectory, DriverStation.getAlliance());
    } else {
      transformedTrajectory = trajectory;
    }

    this.prevTime = -1;

    if (logActiveTrajectory != null) {
      logActiveTrajectory.accept(transformedTrajectory);
    }

    PathPlannerTrajectory.PathPlannerState initialState = transformedTrajectory.getInitialState();

    this.prevSpeeds =
        this.kinematics.toWheelSpeeds(
            new ChassisSpeeds(
                initialState.velocityMetersPerSecond,
                0,
                initialState.curvatureRadPerMeter * initialState.velocityMetersPerSecond));

    this.timer.reset();
    this.timer.start();

    if (this.usePID) {
      this.leftController.reset();
      this.rightController.reset();
    }

    PathPlannerServer.sendActivePath(transformedTrajectory.getStates());
  }

  @Override
  public void execute() {
    double currentTime = this.timer.get();
    double dt = currentTime - this.prevTime;

    if (this.prevTime < 0) {
      this.prevTime = currentTime;
      return;
    }

    Pose2d currentPose = this.poseSupplier.get();
    PathPlannerTrajectory.PathPlannerState desiredState =
        (PathPlannerTrajectory.PathPlannerState) transformedTrajectory.sample(currentTime);

    PathPlannerServer.sendPathFollowingData(desiredState.poseMeters, currentPose);

    ChassisSpeeds targetChassisSpeeds = this.controller.calculate(currentPose, desiredState);
    DifferentialDriveWheelSpeeds targetWheelSpeeds =
        this.kinematics.toWheelSpeeds(targetChassisSpeeds);

    double leftSpeedSetpoint = targetWheelSpeeds.leftMetersPerSecond;
    double rightSpeedSetpoint = targetWheelSpeeds.rightMetersPerSecond;

    double leftOutput;
    double rightOutput;

    if (this.usePID) {
      double leftFeedforward =
          this.feedforward.calculate(
              leftSpeedSetpoint, (leftSpeedSetpoint - this.prevSpeeds.leftMetersPerSecond) / dt);
      double rightFeedforward =
          this.feedforward.calculate(
              rightSpeedSetpoint, (rightSpeedSetpoint - this.prevSpeeds.rightMetersPerSecond) / dt);

      leftOutput =
          leftFeedforward
              + this.leftController.calculate(
                  this.speedsSupplier.get().leftMetersPerSecond, leftSpeedSetpoint);
      rightOutput =
          rightFeedforward
              + this.rightController.calculate(
                  this.speedsSupplier.get().rightMetersPerSecond, rightSpeedSetpoint);
    } else {
      leftOutput = leftSpeedSetpoint;
      rightOutput = rightSpeedSetpoint;
    }

    this.output.accept(leftOutput, rightOutput);
    this.prevSpeeds = targetWheelSpeeds;
    this.prevTime = currentTime;

    if (logTargetPose != null) {
      logTargetPose.accept(desiredState.poseMeters);
    }

    if (logError != null) {
      logError.accept(
          currentPose.getTranslation().minus(desiredState.poseMeters.getTranslation()),
          currentPose.getRotation().minus(desiredState.poseMeters.getRotation()));
    }

    if (logSetpoint != null) {
      logSetpoint.accept(targetChassisSpeeds);
    }
  }

  @Override
  public void end(boolean interrupted) {
    this.timer.stop();

    if (interrupted
        || Math.abs(transformedTrajectory.getEndState().velocityMetersPerSecond) < 0.1) {
      this.output.accept(0.0, 0.0);
    }
  }

  @Override
  public boolean isFinished() {
    return this.timer.hasElapsed(transformedTrajectory.getTotalTimeSeconds());
  }

  private static void defaultLogError(Translation2d translationError, Rotation2d rotationError) {
    SmartDashboard.putNumber("PPRamseteCommand/xErrorMeters", translationError.getX());
    SmartDashboard.putNumber("PPRamseteCommand/yErrorMeters", translationError.getY());
    SmartDashboard.putNumber("PPRamseteCommand/rotationErrorDegrees", rotationError.getDegrees());
  }

  /**
   * Set custom logging callbacks for this command to use instead of the default configuration of
   * pushing values to SmartDashboard
   *
   * @param logActiveTrajectory Consumer that accepts a PathPlannerTrajectory representing the
   *     active path. This will be called whenever a PPRamseteCommand starts
   * @param logTargetPose Consumer that accepts a Pose2d representing the target pose while path
   *     following
   * @param logSetpoint Consumer that accepts a ChassisSpeeds object representing the setpoint
   *     speeds
   * @param logError BiConsumer that accepts a Translation2d and Rotation2d representing the error
   *     while path following
   */
  public static void setLoggingCallbacks(
      Consumer<PathPlannerTrajectory> logActiveTrajectory,
      Consumer<Pose2d> logTargetPose,
      Consumer<ChassisSpeeds> logSetpoint,
      BiConsumer<Translation2d, Rotation2d> logError) {
    PPRamseteCommand.logActiveTrajectory = logActiveTrajectory;
    PPRamseteCommand.logTargetPose = logTargetPose;
    PPRamseteCommand.logSetpoint = logSetpoint;
    PPRamseteCommand.logError = logError;
  }
}

File: pathplannerlib/src/main/native/cpp/pathplanner/lib/commands/PPRamseteCommand.cpp
#include "pathplanner/lib/commands/PPRamseteCommand.h"

#include <utility>
#include <frc/smartdashboard/SmartDashboard.h>
#include <frc/DriverStation.h>

using namespace pathplanner;

std::function<void(PathPlannerTrajectory)> PPRamseteCommand::logActiveTrajectory =
		[](auto traj) {
		};
std::function<void(frc::Pose2d)> PPRamseteCommand::logTargetPose = [](
		auto pose) {
};
std::function<void(frc::ChassisSpeeds)> PPRamseteCommand::logSetpoint = [](
		auto speeds) {
};
std::function<void(frc::Translation2d, frc::Rotation2d)> PPRamseteCommand::logError =
		[](auto transError, auto rotError) {
			frc::SmartDashboard::PutNumber("PPRamseteCommand/xErrorMeters",
					transError.X()());
			frc::SmartDashboard::PutNumber("PPRamseteCommand/yErrorMeters",
					transError.Y()());
			frc::SmartDashboard::PutNumber(
					"PPRamseteCommand/rotationErrorDegrees",
					rotError.Degrees()());
		};

PPRamseteCommand::PPRamseteCommand(PathPlannerTrajectory trajectory,
		std::function<frc::Pose2d()> pose, frc::RamseteController controller,
		frc::SimpleMotorFeedforward<units::meters> feedforward,
		frc::DifferentialDriveKinematics kinematics,
		std::function<frc::DifferentialDriveWheelSpeeds()> wheelSpeeds,
		frc2::PIDController leftController, frc2::PIDController rightController,
		std::function<void(units::volt_t, units::volt_t)> output,
		std::initializer_list<frc2::Subsystem*> requirements,
		bool useAllianceColor) : m_trajectory(std::move(trajectory)), m_pose(
		std::move(pose)), m_controller(controller), m_feedforward(feedforward), m_kinematics(
		kinematics), m_speeds(std::move(wheelSpeeds)), m_leftController(
		std::make_unique < frc2::PIDController > (leftController)), m_rightController(
		std::make_unique < frc2::PIDController > (rightController)), m_outputVolts(
		std::move(output)), m_usePID(true), m_useAllianceColor(useAllianceColor) {
	AddRequirements(requirements);

	if (m_useAllianceColor && m_trajectory.fromGUI
			&& m_trajectory.getInitialPose().X() > 8.27_m) {
		FRC_ReportError(frc::warn::Warning,
				"You have constructed a path following command that will automatically transform path states depending on the alliance color, however, it appears this path was created on the red side of the field instead of the blue side. This is likely an error.");
	}
}

PPRamseteCommand::PPRamseteCommand(PathPlannerTrajectory trajectory,
		std::function<frc::Pose2d()> pose, frc::RamseteController controller,
		frc::SimpleMotorFeedforward<units::meters> feedforward,
		frc::DifferentialDriveKinematics kinematics,
		std::function<frc::DifferentialDriveWheelSpeeds()> wheelSpeeds,
		frc2::PIDController leftController, frc2::PIDController rightController,
		std::function<void(units::volt_t, units::volt_t)> output,
		std::span<frc2::Subsystem* const > requirements, bool useAllianceColor) : m_trajectory(
		std::move(trajectory)), m_pose(std::move(pose)), m_controller(
		controller), m_feedforward(feedforward), m_kinematics(kinematics), m_speeds(
		std::move(wheelSpeeds)), m_leftController(
		std::make_unique < frc2::PIDController > (leftController)), m_rightController(
		std::make_unique < frc2::PIDController > (rightController)), m_outputVolts(
		std::move(output)), m_usePID(true), m_useAllianceColor(useAllianceColor) {
	AddRequirements(requirements);

	if (m_useAllianceColor && m_trajectory.fromGUI
			&& m_trajectory.getInitialPose().X() > 8.27_m) {
		FRC_ReportError(frc::warn::Warning,
				"You have constructed a path following command that will automatically transform path states depending on the alliance color, however, it appears this path was created on the red side of the field instead of the blue side. This is likely an error.");
	}
}

PPRamseteCommand::PPRamseteCommand(PathPlannerTrajectory trajectory,
		std::function<frc::Pose2d()> pose, frc::RamseteController controller,
		frc::DifferentialDriveKinematics kinematics,
		std::function<
				void(units::meters_per_second_t, units::meters_per_second_t)> output,
		std::initializer_list<frc2::Subsystem*> requirements,
		bool useAllianceColor) : m_trajectory(std::move(trajectory)), m_pose(
		std::move(pose)), m_controller(controller), m_kinematics(kinematics), m_outputVel(
		std::move(output)), m_usePID(false), m_useAllianceColor(
		useAllianceColor) {
	AddRequirements(requirements);

	if (m_useAllianceColor && m_trajectory.fromGUI
			&& m_trajectory.getInitialPose().X() > 8.27_m) {
		FRC_ReportError(frc::warn::Warning,
				"You have constructed a path following command that will automatically transform path states depending on the alliance color, however, it appears this path was created on the red side of the field instead of the blue side. This is likely an error.");
	}
}

PPRamseteCommand::PPRamseteCommand(PathPlannerTrajectory trajectory,
		std::function<frc::Pose2d()> pose, frc::RamseteController controller,
		frc::DifferentialDriveKinematics kinematics,
		std::function<
				void(units::meters_per_second_t, units::meters_per_second_t)> output,
		std::span<frc2::Subsystem* const > requirements, bool useAllianceColor) : m_trajectory(
		std::move(trajectory)), m_pose(std::move(pose)), m_controller(
		controller), m_kinematics(kinematics), m_outputVel(std::move(output)), m_usePID(
		false), m_useAllianceColor(useAllianceColor) {
	AddRequirements(requirements);

	if (m_useAllianceColor && m_trajectory.fromGUI
			&& m_trajectory.getInitialPose().X() > 8.27_m) {
		FRC_ReportError(frc::warn::Warning,
				"You have constructed a path following command that will automatically transform path states depending on the alliance color, however, it appears this path was created on the red side of the field instead of the blue side. This is likely an error.");
	}
}

void PPRamseteCommand::Initialize() {
	if (m_useAllianceColor && m_trajectory.fromGUI) {
		m_transformedTrajectory =
				PathPlannerTrajectory::transformTrajectoryForAlliance(
						m_trajectory, frc::DriverStation::GetAlliance());
	} else {
		m_transformedTrajectory = m_trajectory;
	}

	m_prevTime = -1_s;
	PathPlannerTrajectory::PathPlannerState initialState =
			m_transformedTrajectory.sample(0_s);

	m_prevSpeeds = m_kinematics.ToWheelSpeeds(
			frc::ChassisSpeeds { initialState.velocity, 0_mps,
					initialState.velocity * initialState.curvature });
	m_timer.Reset();
	m_timer.Start();
	if (m_usePID) {
		m_leftController->Reset();
		m_rightController->Reset();
	}

	if (PPRamseteCommand::logActiveTrajectory) {
		PPRamseteCommand::logActiveTrajectory (m_transformedTrajectory);
	}
}

void PPRamseteCommand::Execute() {
	auto curTime = m_timer.Get();
	auto dt = curTime - m_prevTime;

	if (m_prevTime < 0_s) {
		if (m_usePID) {
			m_outputVolts(0_V, 0_V);
		} else {
			m_outputVel(0_mps, 0_mps);
		}

		m_prevTime = curTime;
		return;
	}

	PathPlannerTrajectory::PathPlannerState desiredState =
			m_transformedTrajectory.sample(curTime);

	frc::Pose2d currentPose = m_pose();

	auto targetChassisSpeeds = m_controller.Calculate(currentPose,
			desiredState.asWPILibState());
	auto targetWheelSpeeds = m_kinematics.ToWheelSpeeds(targetChassisSpeeds);

	if (m_usePID) {
		auto leftFeedforward = m_feedforward.Calculate(targetWheelSpeeds.left,
				(targetWheelSpeeds.left - m_prevSpeeds.left) / dt);

		auto rightFeedforward = m_feedforward.Calculate(targetWheelSpeeds.right,
				(targetWheelSpeeds.right - m_prevSpeeds.right) / dt);

		auto leftOutput = units::volt_t { m_leftController->Calculate(
				m_speeds().left.value(), targetWheelSpeeds.left.value()) }
				+ leftFeedforward;

		auto rightOutput = units::volt_t { m_rightController->Calculate(
				m_speeds().right.value(), targetWheelSpeeds.right.value()) }
				+ rightFeedforward;

		m_outputVolts(leftOutput, rightOutput);
	} else {
		m_outputVel(targetWheelSpeeds.left, targetWheelSpeeds.right);
	}
	m_prevSpeeds = targetWheelSpeeds;
	m_prevTime = curTime;

	if (PPRamseteCommand::logTargetPose) {
		PPRamseteCommand::logTargetPose(desiredState.pose);
	}

	if (PPRamseteCommand::logError) {
		PPRamseteCommand::logError(
				currentPose.Translation() - desiredState.pose.Translation(),
				currentPose.Rotation() - desiredState.pose.Rotation());
	}

	if (PPRamseteCommand::logSetpoint) {
		PPRamseteCommand::logSetpoint(targetChassisSpeeds);
	}
}

void PPRamseteCommand::End(bool interrupted) {
	m_timer.Stop();

	if (interrupted
			|| std::abs(m_transformedTrajectory.getEndState().velocity())
					< 0.1) {
		if (m_usePID) {
			m_outputVolts(0_V, 0_V);
		} else {
			m_outputVel(0_mps, 0_mps);
		}
	}
}

bool PPRamseteCommand::IsFinished() {
	return m_timer.HasElapsed(m_transformedTrajectory.getTotalTime());
}
